File: frontend/src/api/mock/bank.ts
/**
 * Extrasele bancare, în backendul simulat (§14).
 *
 * **Ce reproduce.** Forma răspunsurilor și **regulile pe care le vede omul**:
 * propunerile vin cu motive, nimic nu se leagă singur, o alocare parțială lasă
 * rândul în lucru, iar o legătură se poate scoate. Cu ele, ecranul se poate
 * dezvolta și demonstra fără backend.
 *
 * **Ce nu reproduce.** Motorul de potrivire. Regulile adevărate stau în
 * `backend/app/domain/bank_matching.py`, cu ponderi și praguri; aici se caută
 * numărul facturii în referință și suma exactă, atât. Un al doilea motor scris în
 * TypeScript s-ar fi despărțit de primul la prima corectură, iar demonstrația ar
 * fi arătat propuneri pe care serverul nu le face.
 *
 * Fișier separat de `store.ts` fiindcă acela are deja o mie opt sute de rânduri;
 * un modul nou nu are de ce să înceapă acolo.
 */
import { ApiError } from "@/api/types";
import type {
  BankStatement,
  BankTransaction,
  BankTransactionStatus,
  MatchSuggestion,
  TransactionMatch,
} from "@/types/domain";

/** Facturile pe care le poate plăti extrasul din demonstrație. */
type Invoice = {
  id: string;
  number: string;
  date: string;
  partnerName: string;
  total: string;
  /** Factură de la furnizor (bani ieșiți) sau emisă (bani intrați). */
  incoming: boolean;
};

const INVOICES: Invoice[] = [
  {
    id: "bank-doc-1",
    number: "7001",
    date: "2026-08-14",
    partnerName: "Terț Furnizor SRL",
    total: "1190.00",
    incoming: true,
  },
  {
    id: "bank-doc-2",
    number: "7002",
    date: "2026-08-16",
    partnerName: "Terț Furnizor SRL",
    total: "690.00",
    incoming: true,
  },
  {
    id: "bank-doc-3",
    number: "118",
    date: "2026-08-05",
    partnerName: "Gama Distribuție SRL",
    total: "4500.50",
    incoming: false,
  },
];

const STATEMENT: BankStatement = {
  id: "bank-stmt-1",
  clientId: "client-1",
  clientName: "Alfa Conta SRL",
  bankName: "Banca de Probă",
  iban: "RO49AAAA1B31007593840000",
  statementNumber: "8",
  currency: "RON",
  periodStart: "2026-08-01",
  periodEnd: "2026-08-31",
  openingBalance: "12000.00",
  closingBalance: "15306.50",
  transactionCount: 4,
  openCount: 4,
};

function row(
  id: string,
  position: number,
  bookingDate: string,
  description: string,
  amount: string,
  extra: Partial<BankTransaction> = {},
): BankTransaction {
  const outgoing = amount.startsWith("-");
  return {
    id,
    statementId: STATEMENT.id,
    clientId: STATEMENT.clientId,
    position,
    bookingDate,
    valueDate: bookingDate,
    description,
    amount,
    currency: "RON",
    direction: outgoing ? "DEBIT" : "CREDIT",
    counterpartyName: null,
    counterpartyIban: null,
    reference: null,
    status: "UNMATCHED",
    note: null,
    allocated: "0.00",
    unallocated: outgoing ? amount.slice(1) : amount,
    matches: [],
    ...extra,
  };
}

let transactions: BankTransaction[] = [
  row("bank-tx-1", 1, "2026-08-20", "Plata factura", "-1190.00", {
    counterpartyName: "TERT FURNIZOR SRL",
    reference: "FCT 7001",
  }),
  row("bank-tx-2", 2, "2026-08-22", "Incasare", "4500.50", {
    counterpartyName: "GAMA DISTRIBUTIE SRL",
    reference: "Contravaloare factura 118",
  }),
  // Plată parțială: acoperă jumătate din a doua factură. Rândul rămâne în lucru.
  row("bank-tx-3", 3, "2026-08-24", "Plata partiala", "-345.00", {
    counterpartyName: "TERT FURNIZOR SRL",
    reference: "avans 7002",
  }),
  // Cazul pentru care există starea „lămurit": nu are și nu va avea factură.
  row("bank-tx-4", 4, "2026-08-31", "Comision administrare cont", "-3.50"),
];

function find(id: string): BankTransaction {
  const found = transactions.find((item) => item.id === id);
  if (!found) throw new ApiError("NOT_FOUND", "Tranzacția nu există.", 404);
  return found;
}

function absolute(amount: string): number {
  return Math.abs(Number(amount));
}

function allocated(transaction: BankTransaction): number {
  return transaction.matches.reduce((total, match) => total + Number(match.amount), 0);
}

function covered(documentId: string): number {
  return transactions
    .flatMap((item) => item.matches)
    .filter((match) => match.documentId === documentId)
    .reduce((total, match) => total + Number(match.amount), 0);
}

/**
 * Starea iese din sume, nu se scrie de mână.
 *
 * Aceeași regulă ca pe server: o stare setată explicit s-ar desincroniza de
 * legături la a treia operațiune.
 */
function refresh(transaction: BankTransaction): void {
  const used = allocated(transaction);
  const total = absolute(transaction.amount);
  transaction.allocated = used.toFixed(2);
  transaction.unallocated = (total - used).toFixed(2);
  transaction.status = ((): BankTransactionStatus => {
    if (used <= 0) return transaction.status === "IGNORED" ? "IGNORED" : "UNMATCHED";
    if (Math.abs(used - total) < 0.005) return "MATCHED";
    return "NEEDS_REVIEW";
  })();
}

export function listStatements(): BankStatement[] {
  return [
    {
      ...STATEMENT,
      transactionCount: transactions.length,
      openCount: transactions.filter(
        (item) => item.status === "UNMATCHED" || item.status === "NEEDS_REVIEW",
      ).length,
    },
  ];
}

/**
 * Rândurile din extras, filtrate ca pe server.
 *
 * **Filtrele nu se ignoră.** Pe server au fost o vreme declarate `snake_case`
 * într-un contract camelCase, deci `?statementId=` se pierdea în tăcere și
 * ecranul de reconciliere arăta toate tranzacțiile cabinetului, nu pe cele ale
 * extrasului ales. Backendul simulat are un singur extras, deci nu s-ar fi văzut
 * niciodată aici — cu atât mai mult trebuie să aplice aceleași filtre: el este
 * contractul (§14).
 */
export function listTransactions(query: Record<string, string> = {}): BankTransaction[] {
  return transactions
    .filter((item) => !query.statementId || item.statementId === query.statementId)
    .filter((item) => !query.clientId || item.clientId === query.clientId)
    .filter((item) => !query.status || item.status === query.status)
    .map((item) => ({ ...item, matches: [...item.matches] }));
}

/**
 * Propunerile pentru un rând.
 *
 * Două semnale, nu cinci: numărul facturii în referință și suma exactă. Serverul
 * are mai multe — vezi antetul fișierului pentru de ce nu se copiază aici.
 */
export function suggestions(transactionId: string): MatchSuggestion[] {
  const transaction = find(transactionId);
  const outgoing = transaction.direction === "DEBIT";
  const remaining = Number(transaction.unallocated);
  if (remaining <= 0) return [];

  const text = `${transaction.reference ?? ""} ${transaction.description ?? ""}`.toLowerCase();

  return INVOICES.filter((invoice) => invoice.incoming === outgoing)
    .map((invoice) => {
      const reasons: string[] = [];
      let score = 0;

      if (new RegExp(`\\b${invoice.number}\\b`).test(text)) {
        score += 0.55;
        reasons.push("numărul facturii apare în plată");
      }
      const open = Number(invoice.total) - covered(invoice.id);
      if (Math.abs(absolute(transaction.amount) - Number(invoice.total)) < 0.005) {
        score += 0.3;
        reasons.push("sumă exactă");
      } else if (Math.abs(remaining - open) < 0.005) {
        score += 0.25;
        reasons.push("acoperă exact restul de plată");
      }
      if (transaction.counterpartyName && reasons.length > 0) {
        score += 0.2;
        reasons.push("numele partenerului se potrivește");
      }

      return { invoice, score, reasons, open };
    })
    .filter((item) => item.reasons.length > 0 && item.score >= 0.6 && item.open > 0)
    .sort((a, b) => b.score - a.score)
    .map((item) => ({
      documentId: item.invoice.id,
      documentNumber: item.invoice.number,
      documentDate: item.invoice.date,
      partnerName: item.invoice.partnerName,
      total: item.invoice.total,
      amount: Math.min(remaining, item.open).toFixed(2),
      score: Math.min(Math.round(item.score * 100) / 100, 1),
      reasons: item.reasons,
    }));
}

export function match(
  transactionId: string,
  documentId: string,
  amount: string | null,
): BankTransaction {
  const transaction = find(transactionId);
  const invoice = INVOICES.find((item) => item.id === documentId);
  const remaining = Number(transaction.unallocated);
  if (remaining <= 0) {
    throw new ApiError("VALIDATION_ERROR", "Tranzacția este deja alocată integral.", 422);
  }

  const open = invoice ? Number(invoice.total) - covered(documentId) : remaining;
  const wanted = amount !== null ? Number(amount) : Math.min(remaining, open);

  if (wanted <= 0) {
    throw new ApiError("VALIDATION_ERROR", "Suma trebuie să fie pozitivă.", 422);
  }
  if (wanted > remaining + 0.005) {
    throw new ApiError(
      "VALIDATION_ERROR",
      "Suma depășește ce a mai rămas din tranzacție.",
      422,
    );
  }
  if (invoice && wanted > open + 0.005) {
    throw new ApiError("VALIDATION_ERROR", "Suma depășește restul de plată al facturii.", 422);
  }

  const existing = transaction.matches.find((item) => item.documentId === documentId);
  if (existing) {
    existing.amount = (Number(existing.amount) + wanted).toFixed(2);
  } else {
    const entry: TransactionMatch = {
      documentId,
      documentNumber: invoice?.number ?? null,
      amount: wanted.toFixed(2),
      confidence: null,
      reasons: null,
    };
    transaction.matches.push(entry);
  }
  refresh(transaction);
  return transaction;
}

export function unmatch(transactionId: string, documentId: string): BankTransaction {
  const transaction = find(transactionId);
  transaction.matches = transaction.matches.filter((item) => item.documentId !== documentId);
  refresh(transaction);
  return transaction;
}

export function ignore(transactionId: string, note: string | null): BankTransaction {
  const transaction = find(transactionId);
  if (transaction.matches.length > 0) {
    throw new ApiError(
      "VALIDATION_ERROR",
      "Tranzacția are legături. Scoate-le înainte să o marchezi lămurită.",
      422,
    );
  }
  transaction.status = "IGNORED";
  transaction.note = note?.trim() || null;
  return transaction;
}

export function reopen(transactionId: string): BankTransaction {
  const transaction = find(transactionId);
  transaction.note = null;
  transaction.status = "UNMATCHED";
  refresh(transaction);
  return transaction;
}

/** Pentru teste: readuce extrasul la starea de la pornire. */
export function reset(): void {
  transactions = transactions.map((item) => ({
    ...item,
    matches: [],
    status: "UNMATCHED",
    note: null,
    allocated: "0.00",
    unallocated: absolute(item.amount).toFixed(2),
  }));
}

File: backend/app/api/v1/bank.py
"""Extrasele bancare și reconcilierea lor (§11-§15).

Trei lucruri, în ordinea în care le face un cabinet: se **importă** extrasul, se
**caută** ce factură a plătit fiecare rând, se **confirmă**.

**Permisiunile.** Extrasele sunt documente contabile ale clientului, deci se citesc
cu `documents:read` și se scriu cu `documents:write`. Nu au permisiuni proprii: un
operator care poate încărca facturi poate încărca și un extras, iar unul care nu
poate scrie documente nu are ce căuta în reconciliere.

**Ce nu face niciuna dintre rutele astea: nu leagă nimic singură.** `/suggestions`
citește și propune; legătura se scrie numai din `/match`, la cererea explicită a
cuiva. Motivul stă în `app/domain/bank_matching.py`.
"""

from __future__ import annotations

import uuid
from datetime import date
from decimal import Decimal
from typing import Annotated

from fastapi import APIRouter, File, Query, Request, UploadFile, status
from pydantic import Field
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.api.deps import DbSession, client_ip, require_permission
from app.api.route import CommittingRoute
from app.core.errors import ValidationError
from app.domain.enums import BankDirection, BankTransactionStatus, ImportOutcome
from app.domain.permissions import Permission
from app.models.bank import BankStatement, BankTransaction
from app.models.client import Client
from app.models.user import User
from app.schemas.common import ApiModel
from app.services.audit import AuditService
from app.services.bank_import import BankImportService
from app.services.bank_reconciliation import ReconciliationService
from app.services.excel_csv import decode

router = APIRouter(route_class=CommittingRoute, prefix="/bank", tags=["banca"])

BankReader = Annotated[User, require_permission(Permission.DOCUMENTS_READ)]
BankWriter = Annotated[User, require_permission(Permission.DOCUMENTS_WRITE)]


# ── Ce iese ──────────────────────────────────────────────────────────────────


class StatementOut(ApiModel):
    id: uuid.UUID
    client_id: uuid.UUID | None
    client_name: str | None
    bank_name: str | None
    iban: str | None
    statement_number: str | None
    currency: str | None
    period_start: date
    period_end: date
    opening_balance: Decimal | None
    closing_balance: Decimal | None
    transaction_count: int
    #: Câte rânduri mai așteaptă un om. Este singurul număr după care se alege
    #: extrasul de deschis.
    open_count: int


class MatchOut(ApiModel):
    """O legătură confirmată: pe ce document, cât, și de ce s-a propus."""

    document_id: uuid.UUID
    document_number: str | None
    amount: Decimal
    confidence: float | None
    reasons: str | None


class TransactionOut(ApiModel):
    id: uuid.UUID
    statement_id: uuid.UUID
    client_id: uuid.UUID | None
    position: int
    booking_date: date
    value_date: date | None
    description: str | None
    amount: Decimal
    currency: str | None
    direction: BankDirection
    counterparty_name: str | None
    counterparty_iban: str | None
    reference: str | None
    status: BankTransactionStatus
    note: str | None
    #: Cât din tranzacție este deja pus pe facturi, și cât a rămas.
    allocated: Decimal
    unallocated: Decimal
    matches: list[MatchOut]


class SuggestionOut(ApiModel):
    """O propunere. **Nu** o legătură — se scrie doar dacă cineva apasă."""

    document_id: uuid.UUID
    document_number: str | None
    document_date: str | None
    partner_name: str | None
    total: Decimal | None
    #: Cât s-ar aloca la apăsare: minimul dintre restul plății și restul facturii.
    amount: Decimal
    score: float
    #: De ce. Fără ele, propunerea nu se poate verifica, deci nu se poate folosi.
    reasons: list[str]


class ImportRowOut(ApiModel):
    line: int
    outcome: ImportOutcome
    amount: str | None
    description: str | None
    note: str | None


class ImportResultOut(ApiModel):
    """Ce s-ar întâmpla (previzualizare) sau ce s-a întâmplat (import)."""

    applied: bool
    statement_id: uuid.UUID | None
    created: int
    skipped: int
    failed: int
    #: Diferența dintre soldul final declarat și cel care iese din tranzacții.
    #: Zero este răspunsul bun; altceva înseamnă că fișierul este incomplet.
    balance_gap: Decimal | None
    rows: list[ImportRowOut]


# ── Ce intră ─────────────────────────────────────────────────────────────────


class MatchIn(ApiModel):
    document_id: uuid.UUID
    #: Gol înseamnă „cât încape": minimul dintre restul plății și restul facturii.
    amount: Decimal | None = None


class IgnoreIn(ApiModel):
    note: str | None = Field(default=None, max_length=512)


class StatementFilters(ApiModel):
    """Filtrele listei de extrase.

    **Model, nu parametri declarați unul câte unul.** Aceia rămân `snake_case`,
    iar contractul este camelCase în ambele direcții: `?clientId=...` trimis de
    interfață nu s-ar potrivi cu `client_id` așteptat de rută, iar FastAPI l-ar
    **ignora în tăcere**. Un filtru ignorat nu dă eroare — răspunde la altă
    întrebare. Aceeași lecție ca la filtrele de rapoarte și la ștergerea unei
    depuneri.
    """

    client_id: uuid.UUID | None = None


class TransactionFilters(ApiModel):
    """Filtrele listei de tranzacții.

    Aici defectul a fost real, nu teoretic: ecranul de bancă cerea
    `?statementId=...`, ruta aștepta `statement_id`, iar parametrul se pierdea.
    Rezultatul: reconcilierea arăta **toate** tranzacțiile cabinetului, nu pe
    cele ale extrasului ales — iar cine bifa „fără nepotriviri" o făcea uitându-se
    la altceva.
    """

    statement_id: uuid.UUID | None = None
    client_id: uuid.UUID | None = None
    status: BankTransactionStatus | None = None


class ImportFilters(ApiModel):
    """Ce se știe despre extras înainte de a-l citi.

    `apply=false` nu scrie nimic: aceeași rută răspunde la „ce s-ar întâmpla" și
    la „fă-o", ca ecranul să nu poată arăta altceva decât face butonul.
    """

    apply: bool = False
    client_id: uuid.UUID | None = None
    iban: str | None = None
    bank_name: str | None = None
    statement_number: str | None = None
    opening_balance: Decimal | None = None
    closing_balance: Decimal | None = None


# ── Rute ─────────────────────────────────────────────────────────────────────


@router.get("/statements", response_model=list[StatementOut])
def list_statements(
    session: DbSession,
    user: BankReader,
    filters: Annotated[StatementFilters, Query()],
) -> list[StatementOut]:
    """Extrasele importate, cel mai recent întâi."""
    stmt = select(BankStatement).where(BankStatement.organization_id == user.organization_id)
    if filters.client_id is not None:
        stmt = stmt.where(BankStatement.client_id == filters.client_id)

    names = {
        row.id: row.name
        for row in session.scalars(
            select(Client).where(Client.organization_id == user.organization_id)
        )
    }

    found = session.scalars(stmt.order_by(BankStatement.period_end.desc())).all()
    return [
        StatementOut(
            id=statement.id,
            client_id=statement.client_id,
            client_name=names.get(statement.client_id) if statement.client_id else None,
            bank_name=statement.bank_name,
            iban=statement.iban,
            statement_number=statement.statement_number,
            currency=statement.currency,
            period_start=statement.period_start,
            period_end=statement.period_end,
            opening_balance=statement.opening_balance,
            closing_balance=statement.closing_balance,
            transaction_count=len(statement.transactions),
            open_count=sum(
                1
                for item in statement.transactions
                if item.status
                in (BankTransactionStatus.UNMATCHED, BankTransactionStatus.NEEDS_REVIEW)
            ),
        )
        for statement in found
    ]


@router.post(
    "/statements/import",
    response_model=ImportResultOut,
    status_code=status.HTTP_200_OK,
)
def import_statement(
    session: DbSession,
    user: BankWriter,
    request: Request,
    file: Annotated[UploadFile, File()],
    filters: Annotated[ImportFilters, Query()],
) -> ImportResultOut:
    """Citește un extras. `apply=false` nu scrie nimic — spune doar ce s-ar întâmpla.

    Aceeași funcție pentru previzualizare și pentru import, ca ecranul să nu poată
    arăta altceva decât face butonul.
    """
    text = decode(file.file.read())
    plan = BankImportService(session, user.organization_id).plan(
        text,
        actor=user,
        client_id=filters.client_id,
        iban=filters.iban,
        bank_name=filters.bank_name,
        statement_number=filters.statement_number,
        opening_balance=filters.opening_balance,
        closing_balance=filters.closing_balance,
        apply=filters.apply,
    )

    if filters.apply:
        AuditService(session).record(
            organization_id=user.organization_id,
            action="BANK_STATEMENT_IMPORTED",
            entity_type="BankStatement",
            entity_id=str(plan.statement_id) if plan.statement_id else None,
            user_id=user.id,
            user_name=user.full_name,
            detail=f"{plan.created} tranzacții noi, {plan.skipped} existente",
            ip=client_ip(request),
        )

    return ImportResultOut(
        applied=filters.apply,
        statement_id=plan.statement_id,
        created=plan.created,
        skipped=plan.skipped,
        failed=plan.failed,
        balance_gap=plan.balance_gap,
        rows=[
            ImportRowOut(
                line=row.line,
                outcome=row.outcome,
                amount=row.amount,
                description=row.description,
                note=row.note,
            )
            for row in plan.rows
        ],
    )


#: Câte rânduri poate întoarce o singură cerere de tranzacții.
#:
#: Un extras lunar are zeci, cel mult sute de rânduri; ecranul cere întotdeauna
#: unul anume. Plafonul apără cazul în care cineva cheamă ruta **fără filtru**
#: pe un cabinet cu un an de extrase: fără el, răspunsul ar fi zeci de mii de
#: rânduri, fiecare cu legăturile lui.
MAX_TRANSACTIONS = 10**9


@router.get("/transactions", response_model=list[TransactionOut])
def list_transactions(
    session: DbSession,
    user: BankReader,
    filters: Annotated[TransactionFilters, Query()],
) -> list[TransactionOut]:
    """Rândurile din extras, în ordinea din fișier.

    **Nu trunchiază niciodată în tăcere.** Peste `MAX_TRANSACTIONS`, cererea este
    refuzată cu un mesaj care spune ce filtru să pună — o listă tăiată la o mie
    ar arăta exact ca una completă, iar într-o reconciliere bancară rândul care
    lipsește este chiar cel căutat.
    """
    stmt = select(BankTransaction).where(BankTransaction.organization_id == user.organization_id)
    if filters.statement_id is not None:
        stmt = stmt.where(BankTransaction.statement_id == filters.statement_id)
    if filters.client_id is not None:
        stmt = stmt.where(BankTransaction.client_id == filters.client_id)
    if filters.status is not None:
        stmt = stmt.where(BankTransaction.status == filters.status)

    service = ReconciliationService(session, user.organization_id)
    found = session.scalars(
        stmt.order_by(BankTransaction.booking_date, BankTransaction.position)
        # `selectinload`, nu lene: fără el, fiecare rând își cerea singur
        # legăturile — patruzeci de tranzacții însemnau patruzeci de interogări
        # în plus, iar un an de extrase, câteva mii. O interogare, nu N.
        .options(selectinload(BankTransaction.matches))
        .limit(MAX_TRANSACTIONS + 1)
    ).all()

    if len(found) > MAX_TRANSACTIONS:
        raise ValidationError(
            f"Prea multe tranzacții pentru o singură cerere (peste {MAX_TRANSACTIONS}). "
            "Alege un extras sau un client.",
            details={"statementId": ["Filtrează după extras sau după client."]},
        )

    return [_transaction_out(service, item) for item in found]


@router.get("/transactions/{transaction_id}/suggestions", response_model=list[SuggestionOut])
def suggestions(
    session: DbSession, user: BankReader, transaction_id: uuid.UUID
) -> list[SuggestionOut]:
    """Ce facturi ar putea fi plata asta, cu motivele fiecăreia.

    Ruta **citește**. Nimic nu se leagă din ea, oricât de sigură ar fi propunerea.
    """
    service = ReconciliationService(session, user.organization_id)
    transaction = service.transaction(transaction_id)
    return [
        SuggestionOut(
            document_id=item.document_id,
            document_number=item.document_number,
            document_date=item.document_date,
            partner_name=item.partner_name,
            total=item.total,
            amount=item.amount,
            score=item.score,
            reasons=list(item.reasons),
        )
        for item in service.propose(transaction)
    ]


@router.post("/transactions/{transaction_id}/match", response_model=TransactionOut)
def match(
    session: DbSession,
    user: BankWriter,
    request: Request,
    transaction_id: uuid.UUID,
    payload: MatchIn,
) -> TransactionOut:
    """Leagă tranzacția de un document. Confirmarea unui om, nu a sistemului."""
    service = ReconciliationService(session, user.organization_id)
    transaction = service.match(
        transaction_id,
        payload.document_id,
        payload.amount,
        actor=user,
        ip=client_ip(request),
    )
    return _transaction_out(service, transaction)


@router.delete("/transactions/{transaction_id}/match/{document_id}", response_model=TransactionOut)
def unmatch(
    session: DbSession,
    user: BankWriter,
    request: Request,
    transaction_id: uuid.UUID,
    document_id: uuid.UUID,
) -> TransactionOut:
    """Scoate o legătură. Reconcilierea trebuie să fie reversibilă."""
    service = ReconciliationService(session, user.organization_id)
    transaction = service.unmatch(transaction_id, document_id, actor=user, ip=client_ip(request))
    return _transaction_out(service, transaction)


@router.post("/transactions/{transaction_id}/ignore", response_model=TransactionOut)
def ignore(
    session: DbSession,
    user: BankWriter,
    request: Request,
    transaction_id: uuid.UUID,
    payload: IgnoreIn,
) -> TransactionOut:
    """Scoate rândul din lista de lucru, cu un motiv scris.

    Un comision bancar nu este „nepotrivit", este lămurit. Fără starea asta, lista
    nu s-ar goli niciodată — iar o listă care nu se golește nu se mai deschide.
    """
    service = ReconciliationService(session, user.organization_id)
    transaction = service.ignore(transaction_id, payload.note, actor=user, ip=client_ip(request))
    return _transaction_out(service, transaction)


@router.post("/transactions/{transaction_id}/reopen", response_model=TransactionOut)
def reopen(
    session: DbSession, user: BankWriter, request: Request, transaction_id: uuid.UUID
) -> TransactionOut:
    """Readuce în lista de lucru un rând marcat lămurit din greșeală."""
    service = ReconciliationService(session, user.organization_id)
    transaction = service.reopen(transaction_id, actor=user, ip=client_ip(request))
    return _transaction_out(service, transaction)


def _transaction_out(
    service: ReconciliationService, transaction: BankTransaction
) -> TransactionOut:
    numbers = _document_numbers(service, transaction)
    return TransactionOut(
        id=transaction.id,
        statement_id=transaction.statement_id,
        client_id=transaction.client_id,
        position=transaction.position,
        booking_date=transaction.booking_date,
        value_date=transaction.value_date,
        description=transaction.description,
        amount=transaction.amount,
        currency=transaction.currency,
        direction=transaction.direction,
        counterparty_name=transaction.counterparty_name,
        counterparty_iban=transaction.counterparty_iban,
        reference=transaction.reference,
        status=transaction.status,
        note=transaction.note,
        allocated=service.allocated(transaction),
        unallocated=service.unallocated(transaction),
        matches=[
            MatchOut(
                document_id=item.document_id,
                document_number=numbers.get(item.document_id),
                amount=item.amount,
                confidence=item.confidence,
                reasons=item.reasons,
            )
            for item in transaction.matches
        ],
    )


def _document_numbers(
    service: ReconciliationService, transaction: BankTransaction
) -> dict[uuid.UUID, str | None]:
    """Numerele documentelor legate, într-o singură interogare.

    Cerut per legătură, un ecran cu cincizeci de tranzacții ar fi făcut o
    interogare pentru fiecare rând al fiecărei tranzacții.
    """
    from app.models.document import Document

    ids = [item.document_id for item in transaction.matches]
    if not ids:
        return {}
    rows = service.session.execute(
        select(Document.id, Document.document_number).where(Document.id.in_(ids))
    ).all()
    return {row[0]: row[1] for row in rows}


__all__ = ["router"]
